added basic kubernetes manifests to deploy to gke

mirioeggmann opened the pull request:
Changes to be able to deploy the project to GKE.

GKE frontend: http://35.201.78.24/
GKE backend: http://34.96.102.132/

mirioeggmann committed:
added basic kubernetes manifests to deploy to gke

bbortt commented on k8s/deployment-backend.yaml:
```diff
@@ -0,0 +1,44 @@
+            - name: APPL_DATASOURCE_USERNAME
+              value: haubzit
+            - name: APPL_DATASOURCE_PASSWORD
+              value: haubzit
```
:)

bbortt commented on backend/Dockerfile:
```diff
@@ -0,0 +1,4 @@
+FROM openjdk:11
+ARG JAR_FILE=target/*.jar
+COPY ${JAR_FILE} app.jar
+ENTRYPOINT ["java","-jar","/app.jar"]
```
don't run this as root. use `CMD` instead of `ENTRYPOINT` because it is an application. `ADD` instead of `COPY` ("no magic"). also, order matters -> build layers. [source](https://github.com/bbortt/Buessle-App/blob/develop/Dockerfile):
```
FROM openjdk:11-slim

CMD ["java", "-jar", "haubzit-back.jar"]

RUN useradd -ms /bin/bash haubzit

EXPOSE 8080
EXPOSE 8090
EXPOSE 9092

USER haubzit
WORKDIR /home/haubzit

ARG JAR_FILE
ADD ${JAR_FILE} ./haubzit-back.jar
```

mirioeggmann replied:
It actually was just a basic image to try if it works on gke ^^ But yes I will change some of the mentioned parts. The part with the root user, ports and the part with the layers. However I will probably not change the part with copy and entrypoint unless you re-explain me why we should do this.

COPY OR ADD (https://docs.docker.com/develop/develop-images/dockerfile_best-practices/ -> ADD or COPY):

> Although ADD and COPY are functionally similar, generally speaking, COPY is preferred. That’s because it’s more transparent than ADD. COPY only supports the basic copying of local files into the container, while ADD has some features (like local-only tar extraction and remote URL support) that are not immediately obvious. Consequently, the best use for ADD is local tar file auto-extraction into the image, as in ADD rootfs.tar.xz /.

ENTRYPOINT OR CMD (https://spring.io/guides/topicals/spring-boot-docker): Every official spring guide example uses entrypoint 🤔

bbortt replied:
I can agree on `COPY`.
`CMD` is a suggestions and will often be used with executables. e.g. you can `docker run -it bash echo "hello world"` which is equal to the following Dockerfile:
```
FROM bash
CMD [echo, "hello world"]
```
because the `ENTRYPOINT` in `bash` is the bash itself.
however, we're not providing an executable here, we just want to run that .jar *and only* that .jar. so, I think not allowing anyone to override our default command is nice. well, at least not that easy.

bbortt commented on frontend/butcher-front/Dockerfile:
```diff
@@ -0,0 +1,9 @@
+RUN npm run build --prod
+
+FROM nginx:1.15.8-alpine
+COPY --from=builder /usr/src/app/dist/butcher-front/ /usr/share/nginx/html
```
take a look at the `backend/Dockerfile`. same rules appy here.

mirioeggmann replied:
Don't know how to change the user in this image without being a big pain in the ass. Feel free to change it if needed. But I think for some static web resources it's not extremely critical.

bbortt replied:
I've got this somewhere. `alipne` is not that hard. will post it tomorrow.

mirioeggmann replied:
Alpine isn't the problem. It's just this specific nginx-alpine image. But yeah probably you just have to edit some configs like /etc/nginx/nginx.conf.

bbortt replied:
i don't get it. why? 👼

mirioeggmann replied:
Changing it to something like this:
```
FROM nginx:1.15.8-alpine

RUN addgroup -S haubzit && adduser -S haubzit -G haubzit

EXPOSE 80

USER haubzit

COPY --from=builder /usr/src/app/dist/butcher-front/ /usr/share/nginx/html
```
Results in:
```
hlfabric@hlfabric-VirtualBox:~/Documents/git/codebase$ kubectl logs haubzit-frontend-5f7976864b-pkx7c
2020/10/12 20:01:26 [warn] 1#1: the "user" directive makes sense only if the master process runs with super-user privileges, ignored in /etc/nginx/nginx.conf:2
nginx: [warn] the "user" directive makes sense only if the master process runs with super-user privileges, ignored in /etc/nginx/nginx.conf:2
2020/10/12 20:01:26 [emerg] 1#1: mkdir() "/var/cache/nginx/client_temp" failed (13: Permission denied)
nginx: [emerg] mkdir() "/var/cache/nginx/client_temp" failed (13: Permission denied)
```

But according to some posts like this https://stackoverflow.com/questions/42329261/running-nginx-as-non-root-user you can change the /etc/nginx/nginx.conf to the desired user and create+change the ownership of some directories (like /var/cache/nginx/client_temp) that would otherwise be created by root when the container starts. I can try to fix it tomorrow.

bbortt replied:
I just looked the one up I had in mind but recognised it was a plain `node` image. let's skip this for now, as you say, it's probably too much of a pain. althought I have some security conserns.. are you sure, something like "directory traveling" isn't possible? is there something like `.htaccess` for apache?

bbortt commented on k8s/deployment-backend.yaml:
```diff
@@ -0,0 +1,44 @@
+          readinessProbe:
+            httpGet:
+              path: /api/option
+              port: 8080
```
port 8090 and actuator would be a thing.

bbortt commented on k8s/deployment-backend.yaml:
```diff
@@ -0,0 +1,44 @@
+              value: haubzit
+          ports:
+            - name: backend
+              containerPort: 8080
```
missing `Socket.io` port (9092 I suppose).

bbortt commented on k8s/ingress-backend.yaml:
```diff
@@ -0,0 +1,8 @@
+spec:
+  backend:
+    serviceName: haubzit-backend
+    servicePort: 8080
```
missing `Socket.io` port.

bbortt commented on k8s/deployment-postgres.yaml:
```diff
@@ -0,0 +1,37 @@
+            - name: POSTGRES_USER
+              value: haubzit
+            - name: POSTGRES_PASSWORD
+              value: haubzit
```
:)

bbortt reviewed (changes requested):
would also be nice to build the file via maven (`dockerfile-maven` plugin). I know it doesn't apply to the frontend for now, but I think it does make sense to add a parent-pom later!

bbortt commented:
oh and please remove both deployment GitHub actions too ;) as for
> just for testing purposes, will add a github action later

I would recommend Travis-CI. but, let's discuss this.

mirioeggmann committed:
moved the frontend folder from butcher-front to just frontend and improved the two dockerfiles

mirioeggmann commented:
@bbortt I changed the two Dockerfiles and deleted the old github deployment workflows. Furthermore I moved the frontend from frontend/butcher-front to just frontend/. Tomorrow I will add the actions to deploy to the gke cluster. Would it be possible for you to add the Actuator Endpoint for the readinessProbe and the Maven Docker plugin?

bbortt committed:
configure spring-boot actuator endpoint.

mirioeggmann committed:
changed gke project name and temporarily changed actuator port from 8090 to 8080 so the deployment still runs until a way is found for the ingress to check on port 8090

mirioeggmann committed:
cleaned configs

mirioeggmann committed:
fix frontend test build

mirioeggmann committed:
added workflow to deploy to gke

mirioeggmann committed:
fixed service account

mirioeggmann committed:
fixed rollout check

mirioeggmann committed:
fixed pom project finalname

mirioeggmann committed:
added readme

mirioeggmann committed:
merge canary in k8s-configs

mirioeggmann committed:
Merge pull request #74 from yb-hackathon-code-butcher/feature/spring-boot-actuator

configure spring-boot actuator endpoint.
